__tests__/errors.spec.ts

```diff
@@ -0,0 +1,219 @@
+            error: {
+              expected: 'number',
+              got: 'vvvv',
+            },
+          },
+          {
+            path: 'invalidString',
+            error: {
+              expected: 'string',
+              got: 1333,
+            },
+          },
+          {
+            path: 'invalidDate',
+            error: { expected: 'date', got: 'no siema eniu' },
+          },
+        ],
+      },
+    );
+  });
+
+  it('should contain args field', () => {
+    const validator = pipe(
+      object({
+        tooShortString: minStringLength(10)(),
+      })(),
+      validate,
+    );
+    expectToMatchError(
+      () =>
+        validator({
+          tooShortString: 'too short',
+        }),
+      {
+        expected: 'object',
+        got: {
+          tooShortString: 'too short',
+        },
+        errors: [
+          {
+            path: 'tooShortString',
+            error: { expected: 'minStringLength', got: 'too short', args: [10] },
+          },
+        ],
+      },
+    );
+  });
+
+  it('should list nested objects which are undefined', () => {
+    const validator = pipe(
+      object({
+        invalidNumber: number(),
+        nested: object({
+          invalidString: string(),
+          deeper: object({
+            invalidDate: date(),
+          })(),
+        })(),
+      }),
+      validate,
+    );
+
+    expectToMatchError(
+      () =>
+        validator({
+          nested: {
+            deeper: {
+              invalidDate: 'I really like this library!',
+            },
+          },
+        }),
+      {
+        expected: 'object',
+        got: { nested: { deeper: { invalidDate: 'I really like this library!' } } },
+        errors: [
+          {
+            path: 'nested',
+            error: {
+              expected: 'object',
+              got: { deeper: { invalidDate: 'I really like this library!' } },
+              errors: [
+                {
+                  path: 'deeper',
+                  error: {
+                    expected: 'object',
+                    got: { invalidDate: 'I really like this library!' },
+                    errors: [
+                      {
+                        path: 'invalidDate',
+                        error: { expected: 'date', got: 'I really like this library!' },
+                      },
+                    ],
+                  },
+                },
+                { path: 'invalidString', error: { expected: 'string', got: undefined } },
+              ],
+            },
+          },
+          { path: 'invalidNumber', error: { expected: 'number', got: undefined } },
+        ],
+      },
+    );
+  });
+
+  describe('array', () => {
+    it('throws expected error details', () => {
+      const validator = validate(array(string())());
+      expectToMatchError(() => validator(['test string', 1231]), {
+        expected: 'array',
+        got: [
+          'test string',
+          new ValidationError(string(), 1231, {
```

Comment:
This seems to be inconsistent with how we handle objects.

How does Yup or Joi return errors in case of arrays and objects?

Reply:
Hmm array is different than the object 🤔  
You can check how it is handled in react hook form resolvers
https://github.com/react-hook-form/resolvers/blob/master/yup/src/yup.ts
https://github.com/react-hook-form/resolvers/blob/master/joi/src/joi.ts

actually, I inspired by joi. This arrays errors are the same as on main branch - I didn't touch anything. How do you expect it is working? I will check how different libraries handles it